### PR TITLE
EVM-friendly TipSet and ECChain formats

### LINKED ISSUE #150: Add power table commitment (CID) and accessory commitment to Tipset structure
Add a power table commitment and a placeholder for an accessory commitment into the Tipset structure. Both are opaque to GPBFT but must be included in considerations of equality and signing payloads (wait for #149 first).

See FIP PR https://github.com/filecoin-project/FIPs/pull/983

### LINKED ISSUE #166: ECTipset Format Redux
@masih, @Kubuxu, and I were discussing the eventual upgrade path to cross-chain bridging and ran into an issue. Even with a 100 tipset limit:

1. We can still easily reach 26KiB+ of data in a single finality certificate.
2. This data is CBOR, so it's hard to make it a private input to a SNARK (requires parsing CBOR in a snark). I think we're going to be _very_ unhappy if we have to start proving facts about CBOR-encoded data inside a snark.
3. Even without snarks, parsing CBOR is non-trivial and not well supported on most chains.

We initially claimed that this was fine because we don't yet support cross-chain bridging but... we still want an upgrade path towards cross-chain bridging (hence the commitment tree). I hate to re-open this can of worms, but I think we need to commit to a sequence of commitment hashes instead of CBOR values. That is:

```go
type Value = [32]byte // root hash of a merkle tree
type ECChain = []Value // concatenated hashes, we could also make a tree here, but I don't think that's necessary.
```

That way, any future SNARKs can:

1. Given a snark friendly commitment to the power-table from the last instance.
2. Validate the signature over the root commitment (32 bytes).
3. Validate a simple merkle-proof down to facts we're trying to prove.
4. Use those facts (e.g., in a new SNARK).

The one objection we envision is observability. But we can handle _that_ by treating this commitment scheme as a part of the signature scheme (suggested by @Kubuxu). Basically:

1. The GossiPBFT protocols still transmit full tipset information.
2. The signature also includes the merkle-proof proving that the signed commitment commits to the transmitted tipste.
3. The v1 finality certificate format can also include the tipsets/power table directly. It's just that the signature will _only_ cover the root hash.

Importantly, this would let us create small finality certificates that _don't_ require parsing CBOR to understand.

### PR DESCRIPTION
**Pull request by Stebalien**
This patch implements the changes described in
https://github.com/filecoin-project/FIPs/pull/1004.

Unfortunately, go-f3 now needs to be aware of the internal structure of `TipSet` objects as it sends them over the wire in one format, but signs another. This effectively reverts #149.

Specific changes:

1. On the wire (and in finality certificates), `TipSet` objects are now directly cbor-marshaled within the payload instead of first being marshaled to byte strings.
2. Instead of directly covering the entire tipset list with the payload signature:
    1. `TipSet` objects are "marshaled for signing" per the FIP change proposed above.
    2. These marshaled tipset objects are stuffed into a merkle tree.
    3. The merkle tree root is signed (along with the payload's other fields), again according the proposed changes.

fixes #166
fixes #150

### CODE REVIEW

File: gpbft/chain.go
Comment (Stebalien) [inline gpbft/chain.go:24]: Thoughts? The only reason to keep it _this_ way is performance, tbh.

File: gpbft/chain.go
Comment (Stebalien) [inline gpbft/chain.go:108]: I'm returning by pointer here and elsewhere as this structure isn't small.

File: gpbft/chain_test.go
Comment (Stebalien) [inline gpbft/chain_test.go:10]: This test really wasn't doing much.

Comment (Stebalien) [reply gpbft/chain_test.go:10]: I'm happy to add something back to test `IsZero`, `Equals` and the stringer if desired.

File: gpbft/gpbft.go
Comment (Stebalien) [inline gpbft/gpbft.go:117]: TODO: write a test asserting the correct layout and expected offsets.

File: gpbft/chain.go
Comment (Stebalien) [inline gpbft/chain.go:68]: TODO: write test asserting the expected layout and offsets.

File: gpbft/chain.go
Comment (Stebalien) [reply gpbft/chain.go:108]: I'm also happy to bikeshed on this (e.g., take by pointer in more places to reduce the dereferencing, etc.).

**Comment by Stebalien** [comment]
A few TODOs (needs a few more tests) but should otherwise be ready for review.

**Comment by Stebalien** [comment]
Hm. Apparently this is killing performance (probably all the hashing). We may need to make that pluggable (e.g., for testing).

File: gpbft/chain_test.go
Comment (masih) [reply gpbft/chain_test.go:10]: Please feel free to remove the test. 

(not now) At some point we should review the exported `gpbft` package APIs and avoid having multiple ways of doing the same thing. Which I suspect might be the root cause.

**Comment by Stebalien** [comment]
If I look at the total time for `TestRepeat`:

- before - 6.3s
- keccak merkle-tree - 41s
- sha2 merkle-tree - 19s
- sha2 merkle-tree + no tipset hashing - 13.6s

File: gpbft/chain.go
Comment (anorth) [inline gpbft/chain.go:15]: Incomplete comment

File: gpbft/chain.go
Comment (anorth) [inline gpbft/chain.go:20]: Can you add a comment describing this derivation?

File: gpbft/chain.go
Comment (anorth) [reply gpbft/chain.go:24]: Yes the dep made me sad. I still prefer to avoid using if it's not too hard.

File: gpbft/chain.go
Comment (anorth) [inline gpbft/chain.go:33]: Not opaque any more! Move the lines of this comment next to the fields, now they are present.

File: gpbft/chain.go
Comment (anorth) [inline gpbft/chain.go:41]:
```suggestion
	Key?      TipSetKey
```

This will probably often show up as `tipset.TipSet`

File: gpbft/chain.go
Comment (anorth) [inline gpbft/chain.go:59]: Why is this using `cbg`? Just `buf.Write`?

File: gpbft/chain.go
Comment (anorth) [reply gpbft/chain.go:108]: Yeah this interface was mostly pointers before I stripped out the type, if I recall. Should probably have more pointer receivers and parameters in that case.

**Review by anorth — COMMENTED** [review]
Partial review, not done yet

File: gpbft/chain.go
Comment (Stebalien) [reply gpbft/chain.go:24]: Can do.

File: gpbft/chain.go
Comment (Stebalien) [reply gpbft/chain.go:59]: Because https://github.com/filecoin-project/FIPs/blob/b98ae5f0e9c0b033279f29fa5c507e457f826947/FIPS/fip-0086.md?plain=1#L307

I'm actually not sure where the wires got crossed in the FEVM FIP. IMO, it should have either been `CBOR([Cid, Cid, Cid])` or `Raw(Cid || Cid || Cid)`. Instead we got `CBOR(Cid || Cid || Cid)`.

Comment (Stebalien) [reply gpbft/chain.go]: Honestly, I'm tempted to change it to `CBOR([Cid, Cid, Cid])` because then we'd be able to dereference the CIDs. E.g.,`$TIPSET_CID/0` would refer to the CID of the first block. But if we did that, we'd probably (eventually) want a small FIP to make FEVM use the new format.

File: sim/signing/bls.go
Comment (Stebalien) [inline sim/signing/bls.go:51]: I'm _actually_ marshaling here because, in this case, I assume we don't really care about performance.

File: sim/signing/fake.go
Comment (Stebalien) [inline sim/signing/fake.go:89]: This skips _all_ hashing. We'll eventually hash with sha256 when signing, but only doing that once is much faster.

File: gpbft/api.go
Comment (Stebalien) [inline gpbft/api.go:96]: I added this as a new function on the base `Host` because it doesn't really fit on either the `Signer` or the `Verifier`. I can also create a new `SignatureOps` (or something) interface that combines `Signer`, `Verifier`, and `MarshalPayloadForSigning`.

File: gpbft/gpbft.go
Comment (Stebalien) [inline gpbft/gpbft.go:408]: This turned out to be "not terrible" because we already needed the host for the network name.

File: gpbft/participant_test.go
Comment (Stebalien) [inline gpbft/participant_test.go:90]: Is this the right way to do this? Basically:

1. We may _not_ call `NetworkName` now, because that's handled inside of `Host.MarshalPayloadForSigning`.
2. Whenever we call `Sign`, we call `Host.MarshalPayloadForSigning`.

File: gpbft/api.go
Comment (Stebalien) [reply gpbft/api.go]: I merged it into a `Signatures` interface. Open to suggestions on better names.

Comment (masih) [reply gpbft/api.go]: Bikeshedding on name; alternative suggestions: 
* `Signing`, 
* `SignerVerifier`
* (don't hate me) separate all the sign related stuff into their own package called `signing` and call the interface `Scheme` then in `gpbft` it would read `signing.Scheme

Comment (Stebalien) [reply gpbft/api.go]: I'll leave 3 for a future patch. Between 1, 2, and the current name... I don't have any preference.

File: merkle/merkle_test.go
Comment (anorth) [inline merkle/merkle_test.go:26]: Add some test for VerifyProof where it should return false.

**Review by anorth — APPROVED** [review]
Looks solid. I suggest waiting for one more peer review on this one, though.

File: merkle/merkle_test.go
Comment (Stebalien) [reply merkle/merkle_test.go:26]: Done.

**Review by masih — APPROVED** [review]

**Review by Kubuxu — COMMENTED** [review]
Late but LGTM

**Pull request merged**
